include/finufft/heuristics.hpp

```diff
@@ -0,0 +1,185 @@
+      if (dim == 3) {
+        if (density >= 16 && epsilon <= 1.0e-3) {
+          return 2.0;
+        }
+        if (density >= 2 && epsilon <= 1.0e-5) {
+          return 2.0;
+        }
+        if (density >= 1 && epsilon <= 1.0e-7) {
+          return 2.0;
+        }
+        if (epsilon <= 1.0e-8) { // for epsilon <= 1.0e-8, 2.0 is favored in 3D always
+          return 2.0;
+        }
+      } // end dim == 3
+    } // end nufftType == 1
+    if (nufftType == 2) {
+      if (dim == 1) {
+        if (density >= 4 && epsilon <= 1.0e-9) {
+          return 2.0;
+        }
+      } // end dim == 1
+      if (dim == 2) {
+        if (density >= 32 && epsilon <= 1.0e-7) {
+          return 2.0;
+        }
+      } // end dim == 2
+      if (dim == 3) {
+        if (density >= 8 && epsilon <= 1.0e-3) {
+          return 2.0;
+        }
+        if (density >= 4 && epsilon <= 1.0e-4) {
+          return 2.0;
+        }
+        if (density >= 1 && epsilon <= 1.0e-5) {
+          return 2.0;
+        }
+        if (density >= 0.5 && epsilon <= 1.0e-7) {
+          return 2.0;
+        }
+        if (density >= 0.25 && epsilon <= 1.0e-8) {
+          return 2.0;
+        }
+      } // end dim == 3
+    } // end nufftType == 2
+  }
+
+  return 1.25;
+}
+
+template<typename T>
+static double bestUpsamplingFactorMultithread(const double density, const int dim,
+                                              const int nufftType, const double epsilon) {
+  constexpr bool isFloat = std::is_same_v<T, float>;
+  if constexpr (isFloat) {
+    // in float multi-threaded, this is the only case where 2.0 is faster
+    if (dim == 3 && nufftType == 1 && epsilon <= 1.0e-6 && density >= 32) {
+      return 2.0;
+    }
+  } else {
+    // In not 3D, 1.25 is always better
+    if (dim != 3) return 1.25;
+
+    if (nufftType == 1) {
+      if (density >= 16 && epsilon <= 1.0e-5) {
+        return 2.0;
+      }
+      if (density >= 8 && epsilon <= 1.0e-7) {
+        return 2.0;
+      }
+      if (density >= 2 && epsilon <= 1.0e-8) {
+        return 2.0;
+      }
+      if (density >= 1 && epsilon <= 1.0e-9) {
+        return 2.0;
+      }
+    }
+    if (nufftType == 2) {
+      if (density >= 16 && epsilon <= 1.0e-5) {
+        return 2.0;
+      }
+      if (density >= 8 && epsilon <= 1.0e-6) {
+        return 2.0;
+      }
+      if (density >= 4 && epsilon <= 1.0e-7) {
+        return 2.0;
+      }
+      if (density >= 2 && epsilon <= 1.0e-8) {
+        return 2.0;
+      }
+      if (density >= 1 && epsilon <= 1.0e-9) {
+        return 2.0;
+      }
+    }
+  }
+
+  return 1.25;
+}
+
+template<typename T>
+double bestUpsamplingFactor(const int nthreads, const double density, const int dim,
+                            const int nufftType, const double epsilon) {
+  // 1) For epsilons <= 1e-9, 1.25 is not supported
+  if (epsilon <= 1.0e-9) {
+    return 2.0;
+  }
+
+  // 2) Special-case for nufftType == 3
+  //    TODO: maybe use the bandwidth here?
```

Comment:
good - let's do it

Reply:
we should do it in setpts where nf is computed already for FFT sizes. Then choose 2.0 if FFT size will be small.

Reply:
Postpone to another PR